smart-contracts/contracts/PublicLock.sol

```diff
@@ -599,6 +649,68 @@ contract PublicLock is ILockCore, ERC165, IERC721, IERC721Receiver, Ownable {
     );
   }
 
+  /**
+   * @dev Destroys the user's key and sends a refund based on the amount of time remaining.
+   * This assumes the caller is the owner or is approved to modify the owner's key.
+   * @param _owner The owner of the key to be destroyed.
+   */
+  function _cancelAndRefund(
+    address _owner
+  )
+    internal
+  {
+    Key storage key = keyByOwner[_owner];
+
+    // 1) Calc refund
+    uint refund = _getCancelAndRefundValue(_owner);
+
+    // 2) Destroy token
+    emit Transfer(
+      _owner,
+      address(0),
+      key.tokenId
+    );
+    emit CancelKey(_owner, refund);
+    // expirationTimestamp is a proxy for hasKey, setting this to `now` instead
+    // of 0 so that we can still differentiate hasKey from hasValidKey.
+    key.expirationTimestamp = now;
+
+    // 3) Issue Refund
+    if (refund > 0) {
+      // Security: doing this last to avoid re-entrancy concerns
+      _owner.transfer(refund);
+    }
+  }
+
+  /**
+   * @dev Determines how much of a refund a key owner would receive if they issued
+   * a cancelAndRefund now.  
+   * @param _owner The owner of the key check the refund value for.
+   */
+  function _getCancelAndRefundValue(
+    address _owner
+  )
+    internal
+    view
+    hasValidKey(_owner)
+    returns (uint refund)
+  {
+    Key storage key = keyByOwner[_owner];
+    // Math: safeSub is not required since `hasValidKey` confirms timeRemaining is positive
+    uint timeRemaining = key.expirationTimestamp - now;
+    // Math: using safeMul in case keyPrice or timeRemaining is very large
+    refund = keyPrice.mul(timeRemaining) / expirationDuration;
+    if (cancelRefundPenaltyDenominator > 0) {
+      uint penalty = keyPrice / cancelRefundPenaltyDenominator;
+      if (refund > penalty) {
+        // Math: safeSub is not required since the if confirms this won't underflow
+        refund -= penalty;
+      } else {
+        refund = 0;
+      }
+    }
```

Comment:
My solidity knowledge is limited but I guess that the return is implicit based on `returns (uint refund)`?

Reply:
@julien51 your guess is correct, `refund` is returned implicitly.